chaos-monkey-spring-boot/src/main/java/de/codecentric/spring/boot/chaos/monkey/watcher/outgoing/ChaosMonkeyWebClientPostProcessor.java

```diff
@@ -0,0 +1,28 @@
+package de.codecentric.spring.boot.chaos.monkey.watcher.outgoing;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanPostProcessor;
+import org.springframework.web.reactive.function.client.WebClient;
+
+/** @author Marcel Becker */
+public class ChaosMonkeyWebClientPostProcessor implements BeanPostProcessor {
+
+  private final ChaosMonkeyWebClientWatcher filter;
+
+  public ChaosMonkeyWebClientPostProcessor(final ChaosMonkeyWebClientWatcher filter) {
+    this.filter = filter;
+  }
+
+  @Override
+  public Object postProcessBeforeInitialization(final Object bean, final String beanName)
+      throws BeansException {
+    final Object target;
+    if (bean instanceof WebClient) {
+      // create a copy of WebClient whose settings are replicated from the current WebClient.
+      target = ((WebClient) bean).mutate().filter(filter).build();
```

Comment:
Similarly, use the [WebClientCusomizer](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/web/reactive/function/client/WebClientCustomizer.html) to add your filter

Reply:
Hi, also here we will have a limitation, that just `WebClients` which are created via injected `WebClient.Builder()` will be customized. In this case we do not have the flexibility, like with `RestTemplate` to combine both approaches, because we cannot control to add the `Filter` only once. @WtfJoke any thoughts how to approach this? The BeanPostProcessor approach is not nice, but safe.

Reply:
> The BeanPostProcessor approach is not nice, but safe.

Actually this approach misses all `WebClients` that are created by the injected builder. Since in my projects we don't pass around individual `WebClient` beans, we instead use composition in the services. 

For example we have an `EmailFacade` and a `UserFacade` the each contains their own `WebClient`s that were created from the builder.

I would recommend using the customizer, and if you want to include the bean post processor, make the filter behave like a `OncePerRequestFilter` by setting an attrbitute on the request or response and not firing if it has already fired. (I couldn't find a `WebClient` example for you, but the concept is used in HttpFilters and Micrometer MeterFilters)

Reply:
So did I get this right. If we use BeanPostProcessor all injected Webclients will be assaulted (but not the one by injected builders). And if we use WebclientCustomizer its vice versa.


> The BeanPostProcessor approach is not nice, but safe.

> Actually this approach misses all `WebClients` that are created by the injected builder.

Arent these statements contradicting? Or am I missing a point here? :sweat_smile: 


So the difficultiy is to assault both but dont assault it twice.

Reply:
Hi @checketts,
I understand your issue, I did not consider this approach. After a sync with @WtfJoke, we are going to implement both approaches and use the `OncePerRequestFilter`pattern to avoid multi assaults. Thank you for the feedback.